File: redhawk/src/base/framework/python/ossie/events/Subscriber.py
#
# This file is protected by Copyright. Please refer to the COPYRIGHT file 
# distributed with this source distribution.
# 
# This file is part of REDHAWK core.
# 
# REDHAWK core is free software: you can redistribute it and/or modify it under 
# the terms of the GNU Lesser General Public License as published by the Free 
# Software Foundation, either version 3 of the License, or (at your option) any 
# later version.
# 
# REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS 
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
# 
# You should have received a copy of the GNU Lesser General Public License 
# along with this program.  If not, see http://www.gnu.org/licenses/.
#

import sys
import time
import Queue
import copy
import threading
import logging
import traceback

from omniORB import any, URI, CORBA
from ossie.cf import CF, CF__POA
import CosEventComm__POA
import CosEventChannelAdmin, CosEventChannelAdmin__POA


class Receiver(CosEventComm__POA.PushConsumer):
    def __init__(self):
        self._recv_disconnect = True
        self.logger = logging.getLogger("ossie.events.Subscriber.Receiver")
        self._lock = threading.Lock()
        self._cond = threading.Condition(self._lock)


    def __del__(self):
        self._cond.acquire()
        self._recv_disconnect = True
        self._cond.notifyAll()
        self._cond.release()
    
    def get_disconnect(self):
        return self._recv_disconnect

    def reset(self):
        self._recv_disconnect = False

    def disconnect_push_consumer(self):
        self.logger.debug("Subscriber.Receiver handle disconnect_push_consumer")
        self._cond.acquire()
        self._recv_disconnect = True
        self._cond.notifyAll()
        self._cond.release()
    
    def wait_for_disconnect( self, wait_time=-1.0, retries=-1 ):
        try:
            self._cond.acquire()
            tries = retries
            while self._recv_disconnect == False:
                self.logger.debug("Subscriber.Reciever::wait_for_disconnect.... waiting for disconnect")
                if wait_time and wait_time > -1.0:
                    self._cond.wait( wait_time/1000.0 )
                    if tries == -1:
                        break
                    tries -= 1
                    if tries < 1:
                        self.logger.debug("Subscriber.Reciever::wait_for_disconnect.... out of retries...")
                        break;
                else:
                    self.logger.debug("Subscriber.Reciever::wait_for_disconnect.... wait forever..")
                    self._cond.wait()
        finally:
            self._cond.release()


class DefaultConsumer(Receiver):
    def __init__(self,parent):
        self.parent = parent
        Receiver.__init__(self)

    def push(self, data):
        if self.parent.dataArrivedCB != None:
            self.parent.logger.trace('Received (callback) DATA: ' + str(data))
            self.parent.dataArrivedCB( data )
        else:
            self.parent.logger.trace('Received (queue) DATA: ' + str(data))
            self.parent.events.put(data)
        

class Subscriber:
    def __init__(self, channel, dataArrivedCB=None):
        self.channel = channel
        self.proxy = None
        self.logger = logging.getLogger('ossie.events.Subscriber')
        self.dataArrivedCB=dataArrivedCB
        self.events = Queue.Queue()
        self.my_lock = threading.Lock()
        self.consumer = DefaultConsumer(self)
        self.connect()


    def __del__(self):
        self.logger.debug("Subscriber  DTOR START")
        if self.consumer and self.consumer.get_disconnect() == False:
            self.logger.debug("Subscriber::DTOR  DISCONNECT")
            self.disconnect()

        self.logger.debug("Subscriber::DTOR  DEACTIVATE CONSUMER")
        self.proxy = None
        self.consumer = None
        self.channel = None
                             
        self.logger.debug("Subscriber  DTOR END")

    def setDataArrivedCB(self, newCB=None ):
        self.dataArrivedCB= newCB

    def terminate(self):
        self.logger.debug("Subscriber::terminate START")
        if self.consumer and self.consumer.get_disconnect() == False:
            self.logger.debug("Subscriber::terminate DISCONNECT")
            self.disconnect()

        self.logger.debug("Subscriber::terminate  DEACTIVATE CONSUMER")
        self.proxy = None
        self.consumer = None
        self.channel = None
                             
        self.logger.debug("Subscriber::terminate END")

    def getData(self):
        retval=None
        if self.dataArrivedCB != None:
            return retval;
        
        try:
            tmp = self.events.get(False,.01)
            self.logger.debug('getData: ' + str(tmp))
            retval = any.from_any(tmp)
        except:
            #print traceback.print_exc()
            retval=None
        
        return retval


    def disconnect(self, retries=10, retry_wait=.01):
        if self.channel == None:
            return -1

        retval=0
        if self.proxy:
            retval=-1
            for x in range(retries):
                try:
                    self.logger.debug("Subscriber::disconnect  disconnect_push_supplier")
                    self.proxy.disconnect_push_supplier()
                    retval=0
                    break
                except CORBA.COMM_FAILURE:
                    pass
                time.sleep(retry_wait)

            if self.consumer:
               self.logger.debug("Subscriber::disconnect  Waiting for disconnect")
               self.consumer.wait_for_disconnect(.01,3)
               self.logger.debug("Subscriber::disconnect  recevied  disconnect")
           
        return retval


    def connect(self, retries=10, retry_wait=.01):
        
        if self.channel == None:
            return -1

        if self.proxy == None :
            for x in range(retries):
                try: 
                    admin = self.channel.for_consumers()
                    break
                except CORBA.COMM_FAILURE:
                    pass
                time.sleep(retry_wait)

            for x in range(retries):
                try: 
                    self.proxy = admin.obtain_push_supplier()
                    break
                except CORBA.COMM_FAILURE:
                    pass
            
                time.sleep(retry_wait)

        if self.proxy == None:
             return
        
        self.logger.debug("Subscriber::connect  Obtained Proxy for EventChannel")

        for x in range(retries):
            try: 
                self.proxy.connect_push_consumer( self.consumer._this() )
                self.consumer.reset()
                break
            except CORBA.BAD_PARAM:
                break
            except CORBA.COMM_FAILURE:
                pass
            except CosEventChannelAdmin.AlreadyConnected:
                self.consumer.reset()
                break
                
            time.sleep(retry_wait)

        self.logger.debug("Subscriber::connect  Connected to EventChannel")

File: redhawk/src/base/framework/java/ossie/src/org/ossie/events/Subscriber.java
/*
 * This file is protected by Copyright. Please refer to the COPYRIGHT file
 * distributed with this source distribution.
 *
 * This file is part of REDHAWK core.
 *
 * REDHAWK core is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */
package org.ossie.events;

import java.util.Queue;
import java.util.LinkedList;
import org.apache.log4j.Logger;
import org.omg.CORBA.ORB;
import org.omg.CORBA.Any;
import org.omg.CORBA.*;
import org.omg.CosEventComm.*;
import org.omg.CosEventChannelAdmin.*;
import CF.EventChannelManagerPackage.*;
import org.ossie.properties.AnyUtils;
import java.lang.InterruptedException;
import java.util.concurrent.locks.*;
import java.util.concurrent.TimeUnit;



public class  Subscriber  {

    public abstract class EventSubscriberConsumer extends PushConsumerPOA {
    };

    public interface  DataArrivedListener  {
	public void processData( final Any data );
    };


    public abstract class Receiver extends PushConsumerPOA {
        
        public Receiver() {};

        public boolean   get_disconnect() { return _recv_disconnect; };

        public void   reset() { _recv_disconnect = false; };
      
        public void disconnect_push_consumer () {
            _logger.debug("::disconnect_push_consumer handle disconnect_push_supplier." );
            _lock.lock();
            try{
                _recv_disconnect =  true;
                _cond.signalAll();
            }finally {
                _lock.unlock();
            }
        };


        public void wait_for_disconnect ( int wait_time, int retries )
        {
            int tries=retries;
            _lock.lock();
            try {
                while( _recv_disconnect == false ) {
                    if ( wait_time > -1 ) {
                        _logger.debug("::wait_for_disconnect.. Waiting on disconnect." );
                        boolean ret = false;
                        try {
                            ret= _cond.await( wait_time, TimeUnit.MILLISECONDS );
                        } catch( InterruptedException e ) {
                        }
                        if ( !ret && tries == -1  ) {
                            break;
                        }
                        if ( tries-- < 1 )  break;
                    }
                    else {
                        try {
                            _cond.await();

                        } catch( InterruptedException e ) {
                        }
                    }
                }
            } finally {
                _lock.unlock();
            }
        
            return;
        };

      protected Lock                        _lock = new ReentrantLock();
      protected Condition                   _cond = _lock.newCondition();
      protected boolean                     _recv_disconnect = true;
      protected Logger                      _logger = Logger.getLogger("ossie.events.Subscriber.Receiver");                     

    };


    private class DefaultConsumer extends Receiver  {

	public DefaultConsumer ( Subscriber inParent ) {
	    parent = inParent;
	}

    
	public  void push( final org.omg.CORBA.Any data ) {
            _logger.trace(" received data from event channel ..");
	    if ( parent != null ) {
		// if parent defines a callback
		if ( parent.dataArrivedCB != null ) {
                    // RESOLVE need to allow for Generic callback method to convert data and
                    // push data...
		    try{
                        parent.dataArrivedCB.processData( data );
		    }
		    catch( Throwable e){
		    }
		}
		else {
		    parent.events.add(data);
		}
	    };
	};


	private Subscriber         parent;
  };



    //
    //  Subscriber for an Event Channel
    //
    // @param channel    event channel returned from the EventChannelManager
    // @param pub   interface that is notified when a disconnect occurs
    // @param retries    number of retries to perform when trying to establish  publisher interface
    // @param retry_wait number of millisecs to wait between retries
    public Subscriber( EventChannel       inChannel ) 
	throws OperationNotAllowed
    {
	this(inChannel,null );
    }

    public Subscriber( EventChannel            inChannel,
		       DataArrivedListener     newListener )
	throws OperationNotAllowed
    {
        _init( inChannel, newListener );
    }


    public int getData( Any ret ) {

      int retval=-1;
      try{

        // check if callback method is enable.. it so then return
        if ( dataArrivedCB != null ) return retval;

        // check if data is available
        if ( events.size() < 1 ) return retval;
          
        Any  rawdata =  events.remove();
        ret.read_value(rawdata.create_input_stream(),rawdata.type());
        retval=0;
      }
      catch( Throwable e) {
      }

      return retval;
    }


    public Any getData()  {                                                                                                                                        

      Any  retval=null;
      try{

        // check if callback method is enable.. it so then return
        if ( dataArrivedCB != null ) return retval;

        // check if data is available
        if ( events.size() < 1 ) return retval;

        return events.remove();
      }
      catch( Throwable e) {
      }

      return retval;
    }

    public void setDataArrivedListener( DataArrivedListener newListener ) {
        dataArrivedCB = newListener;
    }


    public void terminate () {

        logger.trace("Subscriber: TERMINATE START" );

        if ( consumer != null ) {
            org.ossie.corba.utils.deactivateObject(consumer);
        }

        disconnect();

        proxy = null;
        consumer = null;

        logger.trace("Subscriber: TERMINATE END." );
    }


    public int disconnect() {
        return this.disconnect(10,10);
    }

    public int disconnect( int retries, int retry_wait ) {

        logger.debug("Subscriber: disconnect ....." );
        int retval = -1;
        int tries = retries;
        if ( proxy != null ) {
            do {
                try {
                    logger.debug("Subscriber: disconnect_push_supplier... ");
                    proxy.disconnect_push_supplier();
                    retval=0;
                    break;
                }
                catch (COMM_FAILURE ex) {
                    logger.warn("Caught COMM_FAILURE Exception,  diconnecting Push Consumer for Subscriber,  Retrying..." );
                }
                if ( retry_wait > 0 ) {
                    try {
                        java.lang.Thread.sleep(retry_wait*1000);
                    } catch (final InterruptedException ex) {
                    }
                }
                tries--;
            } while(tries>0);

            if ( consumer  != null ){
                logger.debug("Subscriber::disconnect Waiting for disconnect ........" );
                consumer.wait_for_disconnect(1,3);
                logger.debug("Subscriber::disconnect received disconnect." );     
            }
            logger.debug("Subscriber: ProxyPushSupplier disconnected." );
        }
        
        return retval;
    }

    public int connect() {
        return this.connect(10,10);
    }

    public int connect( int retries, int retry_wait ) {

        int retval=-1;
	int tries=retries;

	if ( channel == null ) {
	    logger.warn( "Subscriber,  Channel resource not available ");
	    return retval;
	}

        if ( proxy == null ) {
            ConsumerAdmin         admin=null;
            do {
                try {
                    logger.debug( "Subscriber,  Grab admin object...");
                    admin = channel.for_consumers ();
                    break;
                }
                catch (COMM_FAILURE ex) {
                }
                if ( retry_wait > 0 ) {
                    try {
                        java.lang.Thread.sleep(retry_wait*1000);
                    } catch (final InterruptedException ex) {
                    }

                } 
                tries--;
            } while ( tries > 0 );


            if ( admin == null  ) return retval;

            logger.debug( "Subscriber,  Obtained ConsumerAdmin...");
            tries=retries;
            do {
                try {
                    logger.debug( "Subscriber,  Grab push supplier proxy.");
                    proxy = admin.obtain_push_supplier();
                    break;
                }
                catch (COMM_FAILURE ex) {
                }
                if ( retry_wait > 0 ) {
                    try {
                        java.lang.Thread.sleep(retry_wait*1000);
                    } catch (final InterruptedException ex) {
                    }
                }
                tries--;
            } while ( tries > 0 );

            logger.debug( "Subscriber,  Obtained ProxyPushConsumer." );
        }

	if ( proxy == null  ) return retval;

        PushConsumer sptr=null;
	if  ( consumer != null ) {
            try {
                org.omg.PortableServer.POA poa = org.ossie.corba.utils.OrbContext.RootPOA();
                org.omg.CORBA.Object consumer_obj =  poa.servant_to_reference(consumer);
                sptr = PushConsumerHelper.narrow(consumer_obj);
            }
	    catch(org.omg.PortableServer.POAPackage.ServantNotActive  ex) {
                logger.error( "Subscriber, Caught ServantNotActive exception connecting Push Consumer!");
	    }
	    catch(org.omg.PortableServer.POAPackage.WrongPolicy  ex) {
                logger.error( "Subscriber, Caught WrongPolicy exception connecting Push Consumer!");
	    }
	}

	// now attach supplier to the proxy
        tries=retries;
	do {
	    try {
		proxy.connect_push_consumer( sptr );
                if ( consumer  != null ) {
                    logger.debug( "Subscriber, Reset consumer state ....." );
                    consumer.reset();
                }
                logger.debug( "Subscriber, Connected Consumer to EventChannel....." );
                retval=0;
		break;
	    }

	    catch(TypeError ex) {
                logger.error( "Subscriber, Caught TypeError exception connecting Push Consumer!");
		break;
	    }
	    catch(BAD_PARAM ex) {
                logger.error( "Subscriber, Caught BAD_PARAM exception connecting Push Consumer!");
		break;
	    }
	    catch(AlreadyConnected ex) {
                logger.warn( "Subscriber, Push Consumer already connected!");
                consumer.reset();
                retval=0;
		break;
	    }
	    catch(COMM_FAILURE ex) {
                logger.error( "Subscriber, Caught COMM_FAILURE exception "
				   +"connecting Push Supplier! Retrying...");
	    }
	    if ( retry_wait > 0 ) {
               try {
		   java.lang.Thread.sleep(retry_wait*1000);
                } catch (final InterruptedException ex) {
	       }

	    }
	    tries--;
	} while ( tries > 0);


        return retval;

    };


    // handle to the Event Channel ... duplicate the channel so we own this object 
    protected EventChannel                  channel;

    // handle to object that publishes the event to the channel's consumers
    //protected ProxyPushSupplierOperations   proxy;
    protected ProxyPushSupplier             proxy;

    // handle to object that responds to disconnect messages
    protected Receiver                      consumer;

    //
    // Logger object 
    //
    protected Logger                        logger=null;

    //
    // Callback interface when event messages arrive
    //
    protected DataArrivedListener           dataArrivedCB=null;

    
    protected Queue< Any >                  events=null;


    private  void _init( EventChannel inChannel, DataArrivedListener newListener ) 
        throws OperationNotAllowed 
    {
	logger = Logger.getLogger("ossie.events.Subscriber");

        dataArrivedCB = newListener;

        // create queue to hold events
        events = new LinkedList< Any >();

        // if user passes a bad param then throw...
        channel=inChannel;
	if ( inChannel == null ) throw new OperationNotAllowed();

        // create a local consumer object for the event channel
        consumer = new DefaultConsumer(this);
        if ( consumer != null ) {
            org.ossie.corba.utils.activateObject(consumer, null);
        }

	// connect to the event channel for a subscriber pattern
        connect();
    }


};
